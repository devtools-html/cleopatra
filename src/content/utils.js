// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
