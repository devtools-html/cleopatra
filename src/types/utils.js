/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

/*
 * A not-void type
 * https://flow.org/try/#0C4TwDgpgBAcg9sAanAlgEygXigKClAHygDsBXAWwCMIAnPQqAZ2BpWIHN6jK44AbCAENiXEqT59RAQRo1BIADzCQAPlEB5SgCsIAY2ABuHDgBmpYvpRxiUMxYUAVAFywEydCoAUAN0F9SEC4OAJRBUADe9DQQwKQ0Nr7+EEYAvsZ2up4AjMFGGZ4ARMysHAW5puaZ4ba8LgDklII0dVAp5flkEu2VngDaWQA0UABMALrdFp7maBAmbBBoE5nlQA
 */
export type NotVoidOrNull = number | string | boolean | Array<any> | Object;

export type ExtractReturnType = <V>((...args: any[]) => V) => V;

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
// eslint-disable-next-line no-unused-vars
export type IndexedArray<_IndexType, Value> = Array<Value>;

/**
 * This is a utility type that extracts the return type of a function.
 */
export type $ReturnType<Fn> = $Call<ExtractReturnType, Fn>;
