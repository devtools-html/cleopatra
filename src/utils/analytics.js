/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Document Google Analytics API that is used in the project. These definitions
 * can be updated as more features are used.
 *
 * https://developers.google.com/analytics/devguides/collection/analyticsjs/pages
 */
type GAEvent = {|
  hitType: 'event',
  // Specifies the event category. Must not be empty
  eventCategory: string,
  eventAction: string,
  eventLabel?: string,
  eventValue?: number,
|};

type GAPageView = {|
  hitType: 'pageview',
  page: string,
|};

type GATiming = {|
  hitType: 'timing',
  timingCategory: string,
  timingVar: string,
  timingValue: number,
  timingLabel?: string,
|};

type GAPayload = GAEvent | GAPageView | GATiming;

export type GoogleAnalytics = ('send', GAPayload) => {};

export function sendAnalytics(payload: GAPayload) {
  const ga: ?GoogleAnalytics = self.ga;
  if (ga) {
    ga('send', payload);
  }
}
