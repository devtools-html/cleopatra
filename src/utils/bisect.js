// This file originally comes from https://github.com/3rd-Eden/node-bisection/, but has been
// imported and changed to fix some bugs.

/*
    Copyright (c) 2010/2011 Arnout Kazemier,3rd-Eden
    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE. 
*/

/*
    The two functions here, bisectionLeft and bisectionRight, return the index where a new element
    would be inserted, respectively at the left or at the right of elements with the same value. 
*/

// @flow

export function bisectionRight(
  array: number[] | $TypedArray,
  x: number,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x < array[mid]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

export function bisectionLeft(
  array: number[] | $TypedArray,
  x: number,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x <= array[mid]) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

export function bisectionLeftBy<T>(
  array: T[],
  f: (T) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (f(array[mid]) >= 0) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

export function bisectionLeftByKey<T>(
  array: T[],
  x: number,
  toKey: (T) => number,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x <= toKey(array[mid])) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

// This is the same as bisectionLeftByKey but uses string as the key type.
// If you can find a way to make Flow accept a single function that handles
// both string and number keys, please remove this duplication.
export function bisectionLeftByStrKey<T>(
  array: T[],
  x: string,
  toKey: (T) => string,
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (low < high) {
    const mid = (low + high) >> 1;

    if (x <= toKey(array[mid])) {
      high = mid;
    } else {
      low = mid + 1;
    }
  }

  return low;
}

/*
 * bisectLowerBound(), bisectUpperBound(), and bisectEqualRange() are equivalent to
 * std::lower_bound(), std::upper_bound(), and std::equal_range() respectively.
 *
 * bisectLowerBound() returns an index pointing to the first element in the range
 * in which each element is considered *not less than* the given value passed
 * via |aCompare|, or the length of |aContainer| if no such element is found.
 *
 * bisectUpperBound() returns an index pointing to the first element in the range
 * in which each element is considered *greater than* the given value passed
 * via |aCompare|, or the length of |aContainer| if no such element is found.
 *
 * bisectEqualRange() returns a range [first, second) containing all elements are
 * considered equivalent to the given value via |aCompare|.  If you need
 * either the first or last index of the range, bisectLowerBound() or bisectUpperBound(),
 * which is slightly faster than bisectEqualRange(), should suffice.
 *
 * Example (another example is given in TestBinarySearch.cpp):
 *
 *   Vector<const char*> sortedStrings = ...
 *
 *   struct Comparator {
 *     const nsACString& mStr;
 *     explicit Comparator(const nsACString& aStr) : mStr(aStr) {}
 *     int32_t operator()(const char* aVal) const {
 *       return Compare(mStr, nsDependentCString(aVal));
 *     }
 *   };
 *
 *   auto bounds = bisectEqualRange(sortedStrings, 0, sortedStrings.length(),
 *                            Comparator("needle I'm looking for"_ns));
 *   printf("Found the range [%zd %zd)\n", bounds.first(), bounds.second());
 *
 */
export function bisectLowerBound(
  array: number[] | $TypedArray,
  f: (number) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (high !== low) {
    const middle = (low + high) >> 1;
    const result = f(array[middle]);

    // The range returning from bisectLowerBound does include elements
    // equivalent to the given value i.e. f(element) == 0
    if (result >= 0) {
      high = middle;
    } else {
      low = middle + 1;
    }
  }

  return low;
}

export function bisectUpperBound(
  array: number[] | $TypedArray,
  f: (number) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): number {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (high !== low) {
    const middle = (low + high) >> 1;
    const result = f(array[middle]);

    // The range returning from bisectUpperBound does NOT include elements
    // equivalent to the given value i.e. f(element) == 0
    if (result > 0) {
      high = middle;
    } else {
      low = middle + 1;
    }
  }

  return high;
}

export function bisectEqualRange(
  array: number[] | $TypedArray,
  f: (number) => number, // < 0 if arg is before needle, > 0 if after, === 0 if same
  low?: number,
  high?: number
): [number, number] {
  low = low || 0;
  high = high || array.length;

  if (low < 0 || low > array.length || high < 0 || high > array.length) {
    throw new TypeError("low and high must lie within the array's range");
  }

  while (high !== low) {
    const middle = (low + high) >> 1;
    const result = f(array[middle]);

    if (result > 0) {
      high = middle;
    } else if (result < 0) {
      low = middle + 1;
    } else {
      return [
        bisectLowerBound(array, f, low, middle),
        bisectUpperBound(array, f, middle + 1, high),
      ];
    }
  }

  return [low, high];
}
