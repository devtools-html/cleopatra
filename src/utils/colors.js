/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * These are the colors from Photon. They are inlined to provide easy access. If updating
 * please change the CSS variables as well.
 *
 * Firefox Colors v1.0.3
 * https://github.com/FirefoxUX/design-tokens/tree/master/colors
 */

export const MAGENTA_50 = '#ff1ad9';
export const MAGENTA_60 = '#ed00b5';
export const MAGENTA_70 = '#b5007f';
export const MAGENTA_80 = '#7d004f';
export const MAGENTA_90 = '#440027';

export const PURPLE_50 = '#9400ff';
export const PURPLE_60 = '#8000d7';
export const PURPLE_70 = '#6200a4';
export const PURPLE_80 = '#440071';
export const PURPLE_90 = '#25003e';

export const BLUE_40 = '#45a1ff';
export const BLUE_50 = '#0a84ff';
export const BLUE_60 = '#0060df';
export const BLUE_70 = '#003eaa';
export const BLUE_80 = '#002275';
export const BLUE_90 = '#000f40';

export const TEAL_50 = '#00feff';
export const TEAL_60 = '#00c8d7';
export const TEAL_70 = '#008ea4';
export const TEAL_80 = '#005a71';
export const TEAL_90 = '#002d3e';

export const GREEN_50 = '#30e60b';
export const GREEN_60 = '#12bc00';
export const GREEN_70 = '#058b00';
export const GREEN_80 = '#006504';
export const GREEN_90 = '#003706';

export const YELLOW_50 = '#ffe900';
export const YELLOW_60 = '#d7b600';
export const YELLOW_70 = '#a47f00';
export const YELLOW_80 = '#715100';
export const YELLOW_90 = '#3e2800';

export const RED_50 = '#ff0039';
export const RED_60 = '#d70022';
export const RED_70 = '#a4000f';
export const RED_80 = '#5a0002';
export const RED_90 = '#3e0200';

export const ORANGE_50 = '#ff9400';
export const ORANGE_60 = '#d76e00';
export const ORANGE_70 = '#a44900';
export const ORANGE_80 = '#712b00';
export const ORANGE_90 = '#3e1300';

export const GREY_10 = '#f9f9fa';
export const GREY_20 = '#ededf0';
export const GREY_30 = '#d7d7db';
export const GREY_40 = '#b1b1b3';
export const GREY_50 = '#737373';
export const GREY_60 = '#4a4a4f';
export const GREY_70 = '#38383d';
export const GREY_80 = '#2a2a2e';
export const GREY_90 = '#0c0c0d';

export const INK_70 = '#363959';
export const INK_80 = '#202340';
export const INK_90 = '#0f1126';
