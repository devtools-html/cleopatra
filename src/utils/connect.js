/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import * as React from 'react';
import { connect } from 'react-redux';
import type { Dispatch, State } from '../types/store';

type MapStateToProps<OwnProps: Object, StateProps: Object> = (
  state: State,
  ownProps: OwnProps
) => StateProps;

type MapDispatchToProps<OwnProps: Object, DispatchProps: Object> =
  | ((dispatch: Dispatch, ownProps: OwnProps) => DispatchProps)
  | DispatchProps;

type MergeProps<
  StateProps,
  DispatchProps: Object,
  OwnProps: Object,
  Props: Object
> = (
  stateProps: StateProps,
  dispatchProps: DispatchProps,
  ownProps: OwnProps
) => Props;

type ConnectOptions = {
  pure?: boolean,
  areStatesEqual?: boolean,
  areOwnPropsEqual?: boolean,
  areStatePropsEqual?: boolean,
  areMergedPropsEqual?: boolean,
  storeKey?: boolean,
  withRef?: boolean,
};

export type ExplicitConnectOptions<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
> = {
  mapStateToProps?: MapStateToProps<OwnProps, StateProps>,
  mapDispatchToProps?: MapDispatchToProps<OwnProps, DispatchProps>,
  mergeProps?: MergeProps<
    StateProps,
    DispatchProps,
    OwnProps,
    ConnectedProps<OwnProps, StateProps, DispatchProps>
  >,
  options?: ConnectOptions,
  component: React.ComponentType<
    ConnectedProps<OwnProps, StateProps, DispatchProps>
  >,
};

export type ConnectedProps<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
> = $ReadOnly<{|
  ...OwnProps,
  ...StateProps,
  ...DispatchProps,
|}>;

export type ConnectedComponent<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
> =
  | React.ComponentType<ConnectedProps<OwnProps, StateProps, DispatchProps>>
  | React.StatelessFunctionalComponent<
      ConnectedProps<OwnProps, StateProps, DispatchProps>
    >;

/**
 * react-redux's connect function is too polymorphic and problematic. This function
 * is a wrapper to simplify the typing of connect and make it more explicit, and
 * less magical.
 */
export default function explicitConnect<
  OwnProps: Object,
  StateProps: Object,
  DispatchProps: Object
>(
  connectOptions: ExplicitConnectOptions<OwnProps, StateProps, DispatchProps>
): React.ComponentType<OwnProps> {
  const {
    mapStateToProps,
    mapDispatchToProps,
    mergeProps,
    options,
    component,
  } = connectOptions;

  // Opt out of the flow-typed definition of react-redux's connect, and use our own.
  return (connect: any)(
    mapStateToProps,
    mapDispatchToProps,
    mergeProps,
    options
  )(component);
}
