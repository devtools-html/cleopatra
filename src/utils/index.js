/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/**
 * Firefox has issues switching quickly between fill style colors, as the CSS color
 * is fully parsed each time it is set. As a mitigation, provide a class that only
 * switches the color when it's really needed.
 */
export class FastFillStyle {
  _ctx: CanvasRenderingContext2D;
  _previousFillColor: string;

  constructor(ctx: CanvasRenderingContext2D) {
    this._ctx = ctx;
    this._previousFillColor = '';
  }

  set(fillStyle: string) {
    if (fillStyle !== this._previousFillColor) {
      this._ctx.fillStyle = fillStyle;
      this._previousFillColor = fillStyle;
    }
  }
}

/**
 * Perform a simple shallow object equality check.
 */
export function objectShallowEquals(a: Object, b: Object): boolean {
  let aLength = 0;
  let bLength = 0;
  for (const key in a) {
    if (a.hasOwnProperty(key)) {
      aLength++;
      if (a[key] !== b[key]) {
        return false;
      }
    }
  }
  for (const key in b) {
    if (b.hasOwnProperty(key)) {
      bLength++;
    }
  }
  return aLength === bLength;
}
