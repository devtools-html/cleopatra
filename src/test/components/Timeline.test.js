/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import Timeline from '../../components/timeline';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import mockRaf from '../fixtures/mocks/request-animation-frame';
import { getBoundingBox } from '../fixtures/utils';
import ReactDOM from 'react-dom';

import type { Profile } from '../../types/profile';

function _getProfileWithDroppedSamples(): Profile {
  const { profile } = getProfileFromTextSamples(
    // The base thread is 9 samples long.
    '1  2  3  4  5  6  7  8  9',
    // Create a second thread where `x` is when the thread wasn't yet initialized
    // and where e is an empty sample. The profile fixture will be mutated below
    // to follow this.
    `
      x  x  e  e  A  A  A  x  x
                  B  B  B
                  C  C  H
                  D  F  I
                  E  G
    `
  );

  const [thread1, thread2] = profile.threads;

  // Manually choose the timings:
  const sampleStartIndex = 2;
  const sampleEndIndex = 7;
  Object.assign(thread2, {
    processStartupTime: thread2.samples.time[sampleStartIndex],
    registerTime: thread2.samples.time[sampleStartIndex],
    processShutdownTime: thread2.samples.time[sampleEndIndex],
    unregisterTime: null,
  });
  thread1.name = 'Main Thread';
  thread2.name = 'Thread with dropped samples';

  // Remove the samples that contain 'x' and 'e'.
  {
    const samples = thread2.samples;
    for (const key in samples) {
      if (samples.hasOwnProperty(key) && Array.isArray(samples[key])) {
        // Slice just the stacks we care about, simulating a thread that was started
        // later, and with dropped data in its buffer.
        samples[key] = samples[key].slice(4, 7);
      }
    }
  }
  thread2.samples.length = thread2.samples.time.length;

  profile.threads.push(thread2);
  return profile;
}

describe('Timeline', function() {
  beforeEach(() => {
    jest.spyOn(ReactDOM, 'findDOMNode').mockImplementation(() => {
      // findDOMNode uses nominal typing instead of structural (null | Element | Text), so
      // opt out of the type checker for this mock by returning `any`.
      const mockEl = ({
        getBoundingClientRect: () => getBoundingBox(300, 300),
      }: any);
      return mockEl;
    });

    // Mock out the 2d canvas for the loupe view.
    jest
      .spyOn(HTMLCanvasElement.prototype, 'getContext')
      .mockImplementation(() => mockCanvasContext());
  });
  it('renders the header', () => {
    const flushRafCalls = mockRaf();
    window.devicePixelRatio = 1;
    const ctx = mockCanvasContext();
    /**
     * Mock out any created refs for the components with relevant information.
     */
    function createNodeMock(element) {
      // <TimelineCanvas><canvas /></TimelineCanvas>
      if (element.type === 'canvas') {
        return {
          getBoundingClientRect: () => getBoundingBox(200, 300),
          getContext: () => ctx,
          style: {},
        };
      }
      return null;
    }

    const profile = _getProfileWithDroppedSamples();

    const header = renderer.create(
      <Provider store={storeWithProfile(profile)}>
        <Timeline />
      </Provider>,
      { createNodeMock }
    );

    // We need to flush twice since when the first flush is run, it
    // will request more code to be run in later animation frames.
    flushRafCalls();
    flushRafCalls();

    const drawCalls = ctx.__flushDrawLog();

    expect(header.toJSON()).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.devicePixelRatio;
  });
});
