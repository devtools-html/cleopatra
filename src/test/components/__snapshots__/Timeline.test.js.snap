// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline renders the header 1`] = `
<div
  className="timelineSelection"
  onMouseDown={[Function]}
  onMouseMove={[Function]}
>
  <div
    className="timelineRuler"
  >
    <ol
      className="timelineRulerContainer"
    >
      <li
        className="timelineRulerNotch"
        style={
          Object {
            "left": "0px",
          }
        }
      >
        <span
          className="timelineRulerNotchText"
        >
          0.000s
        </span>
      </li>
      <li
        className="timelineRulerNotch"
        style={
          Object {
            "left": "66.66666666666667px",
          }
        }
      >
        <span
          className="timelineRulerNotchText"
        >
          0.002s
        </span>
      </li>
      <li
        className="timelineRulerNotch"
        style={
          Object {
            "left": "133.33333333333334px",
          }
        }
      >
        <span
          className="timelineRulerNotchText"
        >
          0.004s
        </span>
      </li>
      <li
        className="timelineRulerNotch"
        style={
          Object {
            "left": "200px",
          }
        }
      >
        <span
          className="timelineRulerNotchText"
        >
          0.006s
        </span>
      </li>
      <li
        className="timelineRulerNotch"
        style={
          Object {
            "left": "266.6666666666667px",
          }
        }
      >
        <span
          className="timelineRulerNotchText"
        >
          0.008s
        </span>
      </li>
    </ol>
  </div>
  <div
    className="overflowEdgeIndicator timelineOverflowEdgeIndicator"
  >
    <div
      className="overflowEdgeIndicatorEdge topEdge"
    />
    <div
      className="overflowEdgeIndicatorEdge rightEdge"
    />
    <div
      className="overflowEdgeIndicatorEdge bottomEdge"
    />
    <div
      className="overflowEdgeIndicatorEdge leftEdge"
    />
    <div
      className="overflowEdgeIndicatorScrollbox timelineOverflowEdgeIndicatorScrollbox"
      onScroll={[Function]}
    >
      <div
        className="overflowEdgeIndicatorContentsWrapper"
      >
        <ol
          className="timelineThreadList"
          onMouseDown={[Function]}
        >
          <li
            className="timelineTrack"
          >
            <div
              className="timelineTrackRow timelineTrackGlobalRow"
              onClick={[Function]}
            >
              <div
                className="react-contextmenu-wrapper timelineTrackLabel timelineTrackGlobalGrippy"
                onContextMenu={[Function]}
                onMouseDown={[Function]}
                onMouseOut={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                title={null}
              >
                <h1
                  className="timelineTrackName"
                >
                  Process 0
                </h1>
              </div>
              <div
                className="timelineTrackTrack"
              >
                <div
                  className="timelineTrackThreadBlank"
                />
              </div>
            </div>
            <ol
              className="timelineTrackLocalTracks"
              onMouseDown={[Function]}
            >
              <li
                className="timelineTrack timelineTrackLocal"
              >
                <div
                  className="timelineTrackRow timelineTrackLocalRow selected"
                  onClick={[Function]}
                >
                  <div
                    className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
                    onContextMenu={[Function]}
                    onMouseDown={[Function]}
                    onMouseOut={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchStart={[Function]}
                    title="thread: \\"Main Thread\\" (0)
process: \\"default\\" (0)"
                  >
                    <h1
                      className="timelineTrackName"
                    >
                      Main Thread
                    </h1>
                  </div>
                  <div
                    className="timelineTrackTrack"
                  >
                    <div
                      className="timelineTrackThread"
                    >
                      <div
                        className="timelineTrackThreadIntervalMarkerOverview selected"
                      >
                        <canvas
                          className="timelineTracingMarkersCanvas"
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          onMouseOut={[Function]}
                          onMouseUp={[Function]}
                        />
                      </div>
                      <div
                        className="timelineStackGraph"
                      >
                        <canvas
                          className="timelineStackGraphCanvas"
                          onMouseUp={[Function]}
                        />
                      </div>
                      <div
                        className="timelineEmptyThreadIndicator"
                      />
                    </div>
                  </div>
                </div>
              </li>
              <li
                className="timelineTrack timelineTrackLocal"
              >
                <div
                  className="timelineTrackRow timelineTrackLocalRow"
                  onClick={[Function]}
                >
                  <div
                    className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
                    onContextMenu={[Function]}
                    onMouseDown={[Function]}
                    onMouseOut={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchStart={[Function]}
                    title="thread: \\"Thread with dropped samples\\" (0)
process: \\"default\\" (0)"
                  >
                    <h1
                      className="timelineTrackName"
                    >
                      Thread with dropped samples
                    </h1>
                  </div>
                  <div
                    className="timelineTrackTrack"
                  >
                    <div
                      className="timelineTrackThread"
                    >
                      <div
                        className="timelineTrackThreadIntervalMarkerOverview"
                      >
                        <canvas
                          className="timelineTracingMarkersCanvas"
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          onMouseOut={[Function]}
                          onMouseUp={[Function]}
                        />
                      </div>
                      <div
                        className="timelineStackGraph"
                      >
                        <canvas
                          className="timelineStackGraphCanvas"
                          onMouseUp={[Function]}
                        />
                      </div>
                      <div
                        className="timelineEmptyThreadIndicator"
                      >
                        <div
                          className="timelineEmptyThreadIndicatorBlock timelineEmptyThreadIndicatorStartup"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "left": 0,
                              "width": 66.66666666666667,
                            }
                          }
                        />
                        <div
                          className="timelineEmptyThreadIndicatorBlock timelineEmptyThreadIndicatorShutdown"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "right": 0,
                              "width": 66.66666666666667,
                            }
                          }
                        />
                        <div
                          className="timelineEmptyThreadIndicatorBlock timelineEmptyThreadIndicatorEmptyBuffer"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "left": 66.66666666666667,
                              "width": 66.66666666666667,
                            }
                          }
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </li>
              <li
                className="timelineTrack timelineTrackLocal"
              >
                <div
                  className="timelineTrackRow timelineTrackLocalRow"
                  onClick={[Function]}
                >
                  <div
                    className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
                    onContextMenu={[Function]}
                    onMouseDown={[Function]}
                    onMouseOut={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchStart={[Function]}
                    title="thread: \\"Thread with dropped samples\\" (0)
process: \\"default\\" (0)"
                  >
                    <h1
                      className="timelineTrackName"
                    >
                      Thread with dropped samples
                    </h1>
                  </div>
                  <div
                    className="timelineTrackTrack"
                  >
                    <div
                      className="timelineTrackThread"
                    >
                      <div
                        className="timelineTrackThreadIntervalMarkerOverview"
                      >
                        <canvas
                          className="timelineTracingMarkersCanvas"
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          onMouseOut={[Function]}
                          onMouseUp={[Function]}
                        />
                      </div>
                      <div
                        className="timelineStackGraph"
                      >
                        <canvas
                          className="timelineStackGraphCanvas"
                          onMouseUp={[Function]}
                        />
                      </div>
                      <div
                        className="timelineEmptyThreadIndicator"
                      >
                        <div
                          className="timelineEmptyThreadIndicatorBlock timelineEmptyThreadIndicatorStartup"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "left": 0,
                              "width": 66.66666666666667,
                            }
                          }
                        />
                        <div
                          className="timelineEmptyThreadIndicatorBlock timelineEmptyThreadIndicatorShutdown"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "right": 0,
                              "width": 66.66666666666667,
                            }
                          }
                        />
                        <div
                          className="timelineEmptyThreadIndicatorBlock timelineEmptyThreadIndicatorEmptyBuffer"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                          style={
                            Object {
                              "left": 66.66666666666667,
                              "width": 66.66666666666667,
                            }
                          }
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </li>
            </ol>
          </li>
        </ol>
      </div>
    </div>
  </div>
  <div
    className="timelineSelectionHoverLine"
    style={
      Object {
        "left": "0",
        "visibility": "hidden",
      }
    }
  />
</div>
`;

exports[`Timeline renders the header 2`] = `
Array [
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    0,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    22.22222222222222,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    44.44444444444444,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    66.66666666666666,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    88.88888888888889,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    111.11111111111111,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    133.33333333333331,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    155.55555555555554,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "fillRect",
    177.77777777777777,
    NaN,
    22.22222222222222,
    NaN,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    88.88888888888889,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    111.11111111111111,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    133.33333333333331,
    75,
    22.22222222222222,
    225,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    88.88888888888889,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    111.11111111111111,
    0,
    22.22222222222222,
    300,
  ],
  Array [
    "fillRect",
    133.33333333333331,
    75,
    22.22222222222222,
    225,
  ],
  Array [
    "set fillStyle",
    "#003eaa",
  ],
  Array [
    "clearRect",
    0,
    0,
    300,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "clearRect",
    0,
    0,
    300,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "clearRect",
    0,
    0,
    300,
    300,
  ],
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "scale",
    1,
    1,
  ],
]
`;
